crates/forge/src/test_filter.rs

```diff
@@ -43,34 +48,49 @@ impl TestsFilter {
             IgnoredFilter::NotIgnored
         };
 
-        let name_filter = if exact_match {
-            NameFilter::ExactMatch(test_name_filter.unwrap())
+        let mut name_filters = vec![];
+
+        if exact_match {
+            name_filters.push(NameFilter::ExactMatch(test_name_filter.unwrap()));
         } else if let Some(name) = test_name_filter {
-            NameFilter::Match(name)
+            name_filters.push(NameFilter::Match(name));
         } else {
-            NameFilter::All
+            name_filters.push(NameFilter::All);
         };
 
+        if rerun_failed {
+            name_filters.push(NameFilter::RerunFailed);
+        }
+
         Self {
-            name_filter,
+            name_filters,
             ignored_filter,
         }
     }
 
     pub(crate) fn filter_tests(&self, test_crate: CompiledTestCrateRaw) -> CompiledTestCrateRaw {
         let mut cases = test_crate.test_cases;
-
-        cases = match &self.name_filter {
-            NameFilter::All => cases,
-            NameFilter::Match(filter) => cases
-                .into_iter()
-                .filter(|tc| tc.name.contains(filter))
-                .collect(),
-            NameFilter::ExactMatch(name) => {
-                cases.into_iter().filter(|tc| tc.name == *name).collect()
-            }
-        };
-
+        for name_filter in &self.name_filters {
+            cases = match name_filter {
+                NameFilter::All => cases,
+                NameFilter::Match(filter) => cases
+                    .into_iter()
+                    .filter(|tc| tc.name.contains(filter))
+                    .collect(),
+                NameFilter::ExactMatch(name) => {
+                    cases.into_iter().filter(|tc| tc.name == *name).collect()
+                }
+                NameFilter::RerunFailed => {
+                    match get_cached_failed_tests_names().expect("Can't read cached failed tests") {
```

Comment:
We shouldn't panic here, there can be a legitimate reason that we cannot read this file that is not an hole in our logic e.g. insufficient permissions. Just bubble an error wrapped with `context`